include/ensmallen_bits/config.hpp

```diff
@@ -46,7 +46,18 @@
 #if defined(ENS_USE_OPENMP)
   #define ENS_PRAGMA_OMP_PARALLEL _Pragma("omp parallel")
   #define ENS_PRAGMA_OMP_ATOMIC   _Pragma("omp atomic")
+  #define ENS_PRAGMA_OMP_CRITICAL _Pragma("omp critical")
 #else
   #define ENS_PRAGMA_OMP_PARALLEL
   #define ENS_PRAGMA_OMP_ATOMIC
+  #define ENS_PRAGMA_OMP_CRITICAL
+#endif
+
+
+// Define arma_deprecated for deprecated functionality.
+
+#if defined(_MSC_VER)
+  #define ens_deprecated __declspec(deprecated)
+#else
+  #define ens_deprecated __attribute__((__deprecated__))
```

Comment:
`__attribute__` is only available with gcc and clang, so at a minimum a guard is needed here to allow this code only for these compilers.

For sanity, each compiler should be treated separately though.  I suggest treating gcc and clang as two separate compilers.  Last time I checked clang reports that it's gcc 4.2, but in reality its support for attributes is closer to gcc 4.9.

For gcc, the guard can be along these lines:

    #if defined(__GNUG__) && (!defined(__clang__))
      #undef  ens_deprecated
      #define ens_deprecated __attribute__((__deprecated__))
    #endif

For clang, I suggest to use a guard and clang specific code:
    
    #if defined(__clang__) 
      #if __has_attribute(__deprecated__)
      #undef  ens_deprecated
      #define ens_deprecated __attribute__((__deprecated__))
    #endif

The undefines are for good hygiene.

Reply:
You've done way more fighting with strange compiler versions than I have, so I trust your input here.  But... could I trust your input authoritatively by doing `#define ens_deprecated arma_deprecated`?

Reply:
I'd suggest to have a dedicated `ens_deprecated` with its own declarations. The `arma_deprecated` define is internal to Armadillo and subject to change and/or removal.

Though you can certainly copy'n'paste the current declarations from Armadillo into ensmallen.

(Sidenote: at a future point I may clean up Armadillo and bump the minimum C++ version to C++14 or C++17, which have the [`[[deprecated]]`](https://en.cppreference.com/w/cpp/language/attributes/deprecated) statement builtin. This would make `arma_deprecated` redundant and subject to removal.  Keeping compatibility with C++98/C++03 is starting to get annoying; it's kept to allow use on the ancient RHEL 6, and R folks using Windoze with [fossilised tool chains](https://cran.r-project.org/bin/windows/Rtools/)).

Reply:
Actually, it turns out when you filter down the Armadillo detection mechanism it simplifies to basically what I had plus your bits for gcc and clang.  :)

(Sidenote response: C++14 might be too fast, maybe C++11 is better to aim for?  It seems like the "big three" (gcc, clang, and MSVC) all have reasonably complete C++14 support though, so maybe there is no problem there.)